surround/surround/config.py

```diff
@@ -1,429 +1,231 @@
 
-    Example usage::
+    # Configures whether exceptions thrown during pipeline execution are surfaced.
+    surface_exceptions: bool = False
 
-        config = Config()
-        config.read_from_dict({ "debug": True })
-        config.read_config_files(["config.yaml"])
+@dataclass
+class BaseConfig:
+    """
+    Base structured configuration class that should form the base of all custom
+    configuration classes. Contains properties required by all Surround pipelines.
+
+    :cvar Optional[str] project_root: Absolute path to the root of the Surround project.
+    :cvar Optional[str] package_path: Absolute path to the package of the Surround project.
+    :cvar Optional[str] volume_path: Absolute path to the root of the Surround project (formatted for use in Docker volume commands).
+    :cvar Optional[str] input_path: Absolute path to the input folder where data should be loaded from.
+    :cvar Optional[str] model_path: Absolute path to the models folder where models should be loaded from.
+    :cvar Optional[str] output_path: Generated (datetime naming) absolute path to the output folder where output artifacts should be saved.
+    :cvar SurroundConfig surround: Surround specific configuration that dictates how it runs the pipeline.
+    """
 
-        if config["debug"]:
-            # Do debug stuff
+    # Absolute path to the root of the project.
+    project_root: Optional[str] = field(default_factory=get_project_root_or_cwd)
 
-        for key, value in config:
-            # Iterate over all data
+    # Absolute path to the package that will be executed to run the pipeline.
+    package_path: Optional[str] = field(default_factory=lambda: find_package_path(get_project_root_or_cwd()))
 
-    You could then override the above configuration using the systems environment variables,
-    just prefix the var with `SURROUND_` like so::
+    # Absolute path to the root of the project, formatted for use in Docker commands (auto generated via project_root).
+    volume_path: Optional[str] = field(default_factory=lambda: generate_docker_volume_path(get_project_root_or_cwd()))
 
-        SURROUND_DEBUG=False
+    # Absolute path to the inputs folder where data should be loaded from.
+    input_path: Optional[str] = field(default_factory=lambda: os.path.join(get_project_root_or_cwd(), "input"))
 
-    It also supports overriding nested configuration data, for example with the following config::
+    # Absolute path to the models folder where models should be loaded from.
+    model_path: Optional[str] = field(default_factory=lambda: os.path.join(get_project_root_or_cwd(), "models"))
 
-        predict:
-            debug: True
+    # Absolute path to the output folder where outputs from the current run should be placed (timestamped folder).
+    output_path: Optional[str] = field(default_factory=lambda: os.path.join(get_project_root_or_cwd(), "output", str(datetime.now().strftime("%Y-%m-%d-%H-%M-%S"))))
```

Comment:
There are 2 different output directory now, one here (`output/datetime format`) and the other one is from Hydra (`outputs/date/time`). Is this expected?

<img width="276" alt="Screen Shot 2021-08-30 at 12 18 51 pm" src="https://user-images.githubusercontent.com/5424737/131276236-0d9f95aa-43a3-4281-8069-61559ac0767d.png">